Mathlib/Analysis/Calculus/SecondDerivativeTest.lean

```diff
@@ -0,0 +1,97 @@
+/-
+Copyright (c) 2024 Bjørn Kjos-Hanssen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bjørn Kjos-Hanssen, Jireh Loreaux, Floris van Doorn, Eric Wieser
+-/
+import Mathlib.Analysis.Calculus.FirstDerivativeTest
+import Mathlib.Order.Interval.Set.Basic
+import Mathlib.LinearAlgebra.AffineSpace.Ordered
+
+/-!
+# The Second-Derivative Test
+
+We prove the Second-Derivative test from calculus using the First-Derivative test.
+Source: [Wikipedia](https://en.wikipedia.org/wiki/Derivative_test#Proof_of_the_second-derivative_test).
+
+## Main results
+
+* `isLocalMin_of_deriv_deriv_pos`: The second-derivative test, minimum version.
+
+## Tags
+
+derivative test, calculus
+-/
+
+open Set Filter Topology SignType
+
+section SecondDeriv
+
+variable {f : ℝ → ℝ} {x₀ : ℝ}
+
+/-- If the derivative of `f` is positive at a root `x₀` of `f`, then locally the sign of `f x`
+matches `x - x₀`. -/
+lemma eventually_nhdsWithin_sign_eq_of_deriv_pos (hf : deriv f x₀ > 0) (hx : f x₀ = 0) :
+    ∀ᶠ x in 𝓝 x₀, sign (f x) = sign (x - x₀) := by
+  rw [← nhdsNE_sup_pure x₀, eventually_sup]
+  refine ⟨?_, by simpa⟩
+  have h_tendsto := hasDerivAt_iff_tendsto_slope.mp
+    (differentiableAt_of_deriv_ne_zero <| ne_of_gt hf).hasDerivAt
+  filter_upwards [(h_tendsto.eventually <| eventually_gt_nhds hf),
+    self_mem_nhdsWithin] with x hx₀ hx₁
+  rw [mem_compl_iff, mem_singleton_iff, ← Ne.eq_def] at hx₁
+  obtain (hx' | hx') := hx₁.lt_or_lt
+  · rw [sign_neg (neg_of_slope_pos hx' hx₀ hx), sign_neg (sub_neg.mpr hx')]
+  · rw [sign_pos (pos_of_slope_pos hx' hx₀ hx), sign_pos (sub_pos.mpr hx')]
+
+/-- If the derivative of `f` is negative at a root `x₀` of `f`, then locally the sign of `f x`
+matches `x₀ - x`. -/
+lemma eventually_nhdsWithin_sign_eq_of_deriv_neg (hf : deriv f x₀ < 0) (hx : f x₀ = 0) :
+    ∀ᶠ x in 𝓝 x₀, sign (f x) = sign (x₀ - x) := by
+  simpa [Left.sign_neg, -neg_sub, ← neg_sub x₀] using
+    eventually_nhdsWithin_sign_eq_of_deriv_pos
+        (f := (-f ·)) (x₀ := x₀) (by simpa [deriv.neg]) (by simpa)
+
+lemma deriv_neg_left_of_sign_deriv {f : ℝ → ℝ} {x₀ : ℝ}
+    (h₀ : ∀ᶠ (x : ℝ) in 𝓝 x₀, sign (deriv f x) = sign (x - x₀)) :
```

Comment:
I think in this and the next three lemmas, you should generalize the hypothesis to:
```suggestion
    (h₀ : ∀ᶠ (x : ℝ) in 𝓝[≠] x₀, sign (deriv f x) = sign (x - x₀)) :
```
which you can easily prove by switching the `filter_upwards` line to:
```lean
  filter_upwards [nhdsLT_le_nhdsNE _ h₀, self_mem_nhdsWithin] with x hx' (hx : x < x₀)
```
or some variant thereof (e.g.,`nhdsGT` and `x₀ < x`).